8thwall-camera.js

```diff
@@ -5,88 +5,402 @@
+        this.onException = this.onException.bind(this);
+        this.onCameraStatusChange = this.onCameraStatusChange.bind(this);
+    },
+
+    start: async function () {
+        console.log("useCustomUIOverlays", this.useCustomUIOverlays);
+        if (!this.useCustomUIOverlays) {
+            console.log("Initing")
+            OverlaysHandler.init();
+        }
 
+        try {
+            await this.getPermissions();
+        } catch (error) {
+            // User did not grant the camera or motionEvent permissions
+            window.dispatchEvent(new CustomEvent("8thwall-permission-fail", {detail: error}))
+            return;
+        }
+
+        await waitForXR8();
+
+        XR8.XrController.configure({
+            // enableLighting: true,
+            disableWorldTracking: false
+        });
+
+        this.GlTextureRenderer = XR8.GlTextureRenderer.pipelineModule();
         XR8.addCameraPipelineModules([
-            /* Draw the camera feed */
-            XR8.GlTextureRenderer.pipelineModule(),
-            XR8.XrController.pipelineModule(),
-            this,
+            this.GlTextureRenderer, // Draws the camera feed.
+            XR8.XrController.pipelineModule(), // Enables SLAM tracking.
+            this
         ]);
 
-        if(this.camera == 'back') {
-            XR8.run({canvas: Module['canvas'], ownRunLoop: false});
+        const config = {
+            cameraConfig: {
+                direction: XR8.XrConfig.camera().BACK,
+            },
+            canvas: Module.canvas,
+            allowedDevices: XR8.XrConfig.device().ANY,
+            ownRunLoop: false,
+        }
+        XR8.run(config);
+    },
+
+    /**
+     * @param {*} params 
+     * 
+     * private, called by 8thwall
+     */
+    onAttach: function (params) {
+        this.started = true;
+        WL.scene.colorClearEnabled = false;
 
-        } else if(this.camera == 'back') {
-            XR8.XrController.configure({disableWorldTracking: true});
-            XR8.run({canvas: Module['canvas'], ownRunLoop: false, cameraConfig: {
-                direction: XR8.XrConfig.camera().FRONT
-            }});
-        } else {
-            console.error("[8thwall-camera] Invalid camera setting:", this.camera);
+        const gl = Module.ctx;
+
+        const rot = this.object.rotationWorld;
+        const pos = this.object.getTranslationWorld([]);
+        this.position = Array.from(pos);
+        this.rotation = Array.from(rot);
+
+        XR8.XrController.updateCameraProjectionMatrix({
+            origin: { x: pos[0], y: pos[1], z: pos[2] },
+            facing: { x: rot[0], y: rot[1], z: rot[2], w: rot[3] },
+            // cam: { pixelRectWidth: Module.canvas.width, pixelRectHeight: Module.canvas.height, nearClipPlane: 0.01, farClipPlane: 100 }
+        })
+
+        WL.scene.onPreRender.push(() => {
+            gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, null); // <--- Should not be needed after next nightly is released (current 20230110)
```

Comment:
Will still be needed, since it causes some (usually) unnecessary overhead in the engine to bind the default framebuffer even if not needed.